### ISSUE TITLE
CAMERA and residual issues with xcmsSet in here

### ISSUE THREAD

**Issue by redbluewater** [jorainer/XCMS3_FeatureBasedMN#3]
For a while I think I was getting away with calling xcmsSets, but I think there has been a relatively recent shift to not allowing xcmsSet at all. I had XCMS3_Proprocessing.Rmd working last summer (though my notes are not good enough that I saved my sessionInfo, oops. Now the processing at the step where CAMERA is used to generate information needed for GNPS has a step (line 367) that seems to require an xcmsSet, the line is this:

`xsaC <- groupCorr(xsaF, cor_eic_th = 0.6, pval = 0.05, graphMethod = "hcs",
                  calcCiS = TRUE, calcCaS = TRUE, calcIso = FALSE)`

The error I am getting is
`Error in validObject(.Object) : 
  invalid class “xcmsFileSource” object: superclass "characterORconnection" not defined in the environment of the object's class`

I honestly do not know how to fix this, nor do I have a sense of the magnitude of the change required. Any thoughts?

Thanks for any help.
Krista

```
> sessionInfo()
R version 4.0.5 (2021-03-31)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] RColorBrewer_1.1-2

loaded via a namespace (and not attached):
 [1] BiocManager_1.30.12 compiler_4.0.5      htmltools_0.5.1.1   tools_4.0.5         tinytex_0.31       
 [6] rmarkdown_2.7       knitr_1.31          digest_0.6.27       xfun_0.20           rlang_0.4.10       
[11] rJava_0.9-13        evaluate_0.14      
```

**Comment by jorainer** [comment]
Hi Krista, thanks for the detailed description of your problem. In general, the error you get comes if either a) you saved an object (such as an `xcmsSet`) in a previous analysis/version of R and then load it into a workspace with a newer R version where the definition of the `xcmsSet` class was changed or b) if there is some bug somewhere (doesn't really help :)).

I would suggest you first try to update all R packages for this version of R with:

```r
install.packages("BiocManager")
BiocManager::install(version = "3.12")
```

This call should ensure that your R packages are all from the current stable release 3.12. Eventually then also install the most recent (stable) version of `xcms` for this release from github with: `BiocManager::install("sneumann/xcms", ref = "RELEASE_3_12")`. Then please try again if it works. Alternatively, you could try to *update* the object with `xsaF <- updateObject(xsaF)` - that could/should fix point a) above.

Please let me know if it's working (or not).

For the future, we're about to re-implement the feature grouping (in a more modular and customizable way) in the [MsFeatures](https://bioconductor.org/packages/MsFeatures) package - part of the functionality (e.g. feature grouping based on EIC correlation will then go into `xcms`). Isotope and adduct grouping is not yet in there, but it's on my TODO list.

**Comment by redbluewater** [comment]
Hello Johannes,

That worked, and was a far easier fix than I anticipated. I did some updating to my R packages, but I did not know to do the steps you described. Thanks for sharing them.

The only other change I had to make to **XCMS3_Preprocessing.Rmd** was to change line 160 from this:

`filteredMs2Spectra <- featureSpectra(processedData, return.type = "Spectra")`

to this:

`filteredMs2Spectra <- featureSpectra(processedData, return.type = "MSpectra")`

A quick Google search reveals the same question came up at Steffen Neumann's GitHub site (sneumann/xcms issue #534). Should I post another issue on the DorresteinLaboratory GitHub site? I am not sure of the easiest way to propagate this correction across GitHub repositories.

Thanks for your help.
Cheers,
Krista
